test/jit/test_module_containers.py

```diff
@@ -428,3 +428,51 @@ def forward(self, inputs):
 
         m = MyModule()
         self.checkModule(m, [torch.randn(2, 2)])
+
+    def test_typed_module_dict(self):
+        """
+        Test that a type annotation can be provided for a ModuleDict that allows
+        non-static indexing.
+        """
+        @torch.jit.interface
+        class ModuleInterface(torch.nn.Module):
+            def forward(self, inp: Any) -> Any:
+                pass
+
+
+        class MyModule(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+
+            def forward(self, inp: Any) -> Any:
+                if isinstance(inp, torch.Tensor):
+                    return torch.max(inp, dim=0)
+
+                return inp
+
+
+        class Mod(torch.nn.Module):
+            __annotations__ = {"d": Dict[str, ModuleInterface]}
```

Comment:
Thought: I wonder if we'd be better off with an annotation that isn't "Dict". It's not literally a dict. Maybe `Mapping[str, int]` (the built-in mypy type) ? Or just `nn.ModuleDict`. I don't think this is a blocker but something to think about if we were to make this public.